src/typechecker.erl

```diff
@@ -4230,45 +4403,76 @@ describe_expr(_)                          -> "expression".
 
 -spec print_type_error(erl_parse:abstract_expr(),
                        typelib:extended_type(),
-                       typelib:extended_type()) -> ok.
-print_type_error(Expression, ActualType, ExpectedType) ->
-    io:format("The ~s ~s on line ~p~s is expected "
-              "to have type ~s but it has type ~s~n",
-              [describe_expr(Expression),
-               erl_pp:expr(Expression),
-               line_no(Expression),
-               maybe_format_column(Expression),
-               typelib:pp_type(ExpectedType),
-               typelib:pp_type(ActualType)]).
-
-pp_intersection_type([]) ->
+                       typelib:extended_type(),
+                       proplists:proplist()) -> ok.
+print_type_error(Expression, ActualType, ExpectedType, Opts) ->
+    io:format("~sThe ~s ~ts~s is expected "
+              "to have type ~ts but it has type ~ts~n",
+              [format_location(Expression, brief, Opts),
+               describe_expr(Expression),
+               pp_expr(Expression, Opts),
+               format_location(Expression, verbose, Opts),
+               pp_type(ExpectedType, Opts),
+               pp_type(ActualType, Opts)]).
+
+pp_intersection_type([], _) ->
     "";
 %% TODO: pp_type seems to have problems printing bounded types.
-pp_intersection_type([{type, _, bounded_fun, [Ty, []]} | Tys]) ->
-    typelib:pp_type(Ty) ++ ["\n" || Tys /= []] ++ pp_intersection_type(Tys);
-pp_intersection_type([Ty|Tys]) ->
-    typelib:pp_type(Ty) ++ ["\n" || Tys /= []] ++ pp_intersection_type(Tys).
-
+pp_intersection_type([{type, _, bounded_fun, [Ty, []]} | Tys], Opts) ->
+    pp_type(Ty, Opts) ++ ["\n" || Tys /= []] ++ pp_intersection_type(Tys, Opts);
+pp_intersection_type([Ty|Tys], Opts) ->
+    pp_type(Ty, Opts) ++ ["\n" || Tys /= []] ++ pp_intersection_type(Tys, Opts).
+
+pp_expr(Expr, Opts) ->
+    case proplists:get_value(fmt_expr_fun, Opts) of
+        Fun when is_function(Fun) ->
+            Fun(Expr);
+        _ ->
+            erl_pp:expr(Expr)
+    end.
 
+pp_type(Ty, Opts) ->
+    case proplists:get_value(fmt_type_fun, Opts) of
```

Comment:
it was only my subconscious immature optimiser that tried to avoid creating a fun object in the default case. (btw creating such a fun object just got cheaper in OTP 21) But this is not a high-freq trading software and probably your suggestion is better-looking.

Reply:
Ah, I didn't even think about performance. I don't mind, either solution is good to me.